src/webgpu/api/validation/query_set/create.spec.ts

```diff
@@ -2,140 +2,110 @@ export const description = `
-      // invalid
-      ['clipper-invocations'],
-    ]),
-  ])
-  .fn(t => {
-    const { count = 1, pipelineStatistics } = t.params as {
-      count?: number;
-      pipelineStatistics?: GPUPipelineStatisticName[];
-    };
+  .params(
+    params()
+      .combine(poptions('type', kQuerytype))
+      .combine(poptions('count', [0, kMaxQueryCount, kMaxQueryCount + 1]))
+  )
+  .fn(async t => {
+    const { type, count } = t.params;
+
+    if (type === 'pipeline-statistics') {
+      await t.selectDeviceOrSkipTestCase({
+        extensions: ['pipeline-statistics-query'],
+      });
+    } else if (type === 'timestamp') {
+      await t.selectDeviceOrSkipTestCase({
+        extensions: ['timestamp-query'],
+      });
+    }
+
+    const pipelineStatistics =
+      type === 'pipeline-statistics' ? (['clipper-invocations'] as const) : ([] as const);
 
     t.expectValidationError(() => {
-      t.device.createQuerySet({ type: 'occlusion', count, pipelineStatistics });
-    }, count > kMaxQueryCount || (pipelineStatistics !== undefined && pipelineStatistics.length > 0));
+      t.device.createQuerySet({ type, count, pipelineStatistics });
+    }, count > kMaxQueryCount);
   });
 
-g.test('pipeline_statistics')
+g.test('pipelineStatistics')
   .desc(
     `
-Tests that create query set with the type of pipeline-statistics.
-- control cases:
-  - count <= kMaxQueryCount
-  - pipelineStatistics is set with pipeline statistic names
-- invalid cases:
-  - count > kMaxQueryCount
-  - pipelineStatistics is undefined, empty or duplicated
+Tests that create query set with the GPUPipelineStatisticName for all query types:
+- pipelineStatistics is undefined or empty
+- pipelineStatistics is a sequence of valid values
+- pipelineStatistics is a sequence of duplicate values
+- x= {occlusion, pipeline-statistics, timestamp} query
   `
   )
-  .params([
-    ...poptions('count', kQueryCounts),
-    ...poptions('pipelineStatistics', [
-      // valid
-      ['clipper-invocations'],
-      // invalid
-      undefined,
-      [],
-      ['clipper-invocations', 'clipper-invocations'],
-    ]),
-  ])
+  .params(
+    params()
+      .combine(poptions('type', kQuerytype))
+      .combine(
+        poptions('pipelineStatistics', [
+          undefined,
+          [] as const,
+          ['clipper-invocations'] as const,
+          ['clipper-invocations', 'clipper-invocations'] as const,
+          [
+            'clipper-invocations',
+            'clipper-primitives-out',
+            'compute-shader-invocations',
+            'fragment-shader-invocations',
+            'vertex-shader-invocations',
+          ] as const,
+        ])
+      )
+      // Except pipeline statistics query, there is no need to test duplicate values and all values of GPUPipelineStatisticName
+      .unless(
+        ({ type, pipelineStatistics }) =>
+          type !== 'pipeline-statistics' &&
+          pipelineStatistics !== undefined &&
+          pipelineStatistics.length >= 2
+      )
```

Comment:
Thanks, these tests did end up clearer with this split.

Now that I am looking at them this way, this is the last remaining source of complexity. I think we can clarify even further by splitting this test - one for each validation rule. And now that I think about it, that's a very useful guiding rule. What do you think of this split?

- One test for the rule "pipelineStatistics must be [] for non-pipeline-statistics queries": type in {all types} &times; pipelineStatistics in { undefined, [], ['clipper-invocations'] }
- One test for the rule "pipelineStatistics must not have duplicates": type = pipeline-statistics, pipelineStatistics in { ... }

Reply:
If there's a rule that pipelineStatistics must have at least one entry for `'pipeline-statistics`' query sets, that can roll into the second test I suggested.

Reply:
A little unclear. Do you mean split the 'pipelineStatistics' tests to two tests, one for non-pipeline-statistics queries and other for pipeline-statistics query? or spilt the checkpoint of 'pipelineStatistics' to two parts, one is { undefined, [], ['clipper-invocations'] } for all query types, the left is duplicated values and full values for pipeline-statistics query?

Reply:
Sorry it's not that clear, it's hard to explain :)

> spilt the checkpoint of 'pipelineStatistics' to two parts, one is { undefined, [], ['clipper-invocations'] } for all query types, the left is duplicated values and full values for pipeline-statistics query?

This exactly. Message me on slack if you would like to chat more about it; I'll try to be online a bit.

Reply:
Thanks, I got your point.